### PR TITLE
Implementation of symbolic function in solveset

### LINKED ISSUE #12239: solveset() doesn't support symbolic functions yet
Example with SymPy 1.0 with Python 3.6:

```python
from sympy import solveset 
import sympy as sm

x  = sm.Symbol('x')  
f  = sm.Function('f')  

# this works 
sm.solve(f(0)**2 - 1, f(0))

# so does this
sm.nsolve([f(0)**2 - 1], [f(0)], [3])

# this doesn't
solveset(f(1)**2 - 1, f(1))

# but this does
solveset(x**2 - 1, x)
```
The second to last example gives:

```python
NotImplementedError: 
relationship between value and 0 is unknown: Eq(f(1)**2 - 1, 0)
```

So I guess this hasn't been added yet. 

This might be related to another inconsistency in symbolic functions in #12169.

### PR DESCRIPTION
**Pull request by arighnac**
Fixes #12239.

This PR implements symbolic functions in solveset.

**PREVIOUS**

`>>> from sympy import solveset` 
`>>> import sympy as sm`
`>>> x  = sm.Symbol('x')`
`>>> f  = sm.Function('f')`

`>>> solveset(f(1)**2 - 1, f(1))`

gave **output** as -

`NotImplementedError:`
`relationship between value and 0 is unknown: Eq(f(1)**2 - 1, 0)`

**NOW** 

Same piece of code gives **output** as -

`{-1,1}`

### CODE REVIEW

**Comment by ArighnaIITG** [comment]
@smichr , @gxyd , @ylemkimon 
Please review this PR.

File: sympy/solvers/solveset.py [hunk 1]
```diff
@@ -896,6 +897,14 @@ def solveset(f, symbol=None, domain=S.Complexes):
     if not isinstance(f, (Expr, Number)):
         raise ValueError("%s is not a valid SymPy expression" % (f))
 
+    if (isinstance(symbol, Function) and not symbol.has(Symbol)):
```
Comment (ylemkimon) [inline sympy/solvers/solveset.py]: A function can have symbol and still solvable like: `solveset(f(x) - 1, f(x))`

File: sympy/solvers/solveset.py [hunk 2]
```diff
@@ -896,6 +897,14 @@ def solveset(f, symbol=None, domain=S.Complexes):
...
+        for arg in f.args:
+            if not (arg.has(symbol) or arg.is_Number):
+                raise ValueError("%s is not a valid SymPy expression" % (f))
```
Comment (ylemkimon) [inline sympy/solvers/solveset.py]: It doesn't mean that it is not a valid expression, rather it means that it is not solvable using this method and should be passed to other methods.

Comment (ArighnaIITG) [reply sympy/solvers/solveset.py]: How do I implement this? How do I pass it to other solvers? @ylemkimon

Comment (ylemkimon) [reply sympy/solvers/solveset.py]: Do not raise a error or return.

**Comment by Yathartha22** [comment]
The logic that you have applied has already been implemented in #12245 . I am not sure that it would be a good idea to create a new PR(to retain authorship).
@ylemkimon

**Comment by smichr** [comment]
> already been implemented in #12245

Yes, making a PR to that PR would be a good way to combine efforts.

**Comment by ArighnaIITG** [comment]
How do I create a PR to that PR?
Btw, #12245  has some merge conflicts. Will it be good enough to create a PR to a PR which has merge conflicts?

**Pull request closed without merging**